cpp/src/merge/merge.cu

```diff
@@ -0,0 +1,453 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <thrust/execution_policy.h>
+#include <thrust/for_each.h>
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/tuple.h>
+#include <thrust/device_vector.h>
+#include <thrust/merge.h>
+
+#include <algorithm>
+#include <utility>
+#include <vector>
+#include <memory>
+#include <type_traits>
+
+#include <cudf/types.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/table/row_operators.cuh>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <rmm/thrust_rmm_allocator.h>
+#include <utilities/legacy/cuda_utils.hpp>
+#include <cudf/utilities/bit.hpp>
+#include <cudf/null_mask.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+//#include <cudf/strings/detail/merge.cuh> // <- TODO: separate PR for strings support
+
+#include <cudf/detail/merge.hpp>
+
+namespace { // anonym.
+
+using namespace cudf;
+
+using experimental::detail::side;
+using index_type = experimental::detail::index_type;
+
+
+/**
+ * @brief Merges the bits of two validity bitmasks.
+ *
+ * Merges the bits from two column_device_views into the destination column_device_view
+ * according to `merged_indices` map such that bit `i` in `out_col`
+ * will be equal to bit `thrust::get<1>(merged_indices[i])` from `left_dcol`
+ * if `thrust::get<0>(merged_indices[i])` equals `side::LEFT`; otherwise,
+ * from `right_dcol`.
+ *
+ * `left_dcol`, `right_dcol` and `out_dcol` must not
+ * overlap.
+ *
+ * @tparam left_have_valids Indicates whether left_dcol mask is unallocated (hence, ALL_VALID)
+ * @tparam right_have_valids Indicates whether right_dcol mask is unallocated (hence ALL_VALID)
+ * @param[in] left_dcol The left column_device_view whose bits will be merged
+ * @param[in] right_dcol The right column_device_view whose bits will be merged
+ * @param[out] out_dcol The output mutable_column_device_view after merging the left and right
+ * @param[in] num_destination_rows The number of rows in the out_dcol
+ * @param[in] merged_indices The map that indicates the source of the input and index
+ * to be copied to the output. Length must be equal to `num_destination_rows`
+ */
+template <bool left_have_valids, bool right_have_valids>
+__global__ void materialize_merged_bitmask_kernel(column_device_view left_dcol,
+                                                  column_device_view right_dcol,
+                                                  mutable_column_device_view out_dcol,
+                                                  size_type const num_destination_rows,
+                                                  index_type const* const __restrict__ merged_indices) {
+  size_type destination_row = threadIdx.x + blockIdx.x * blockDim.x;
+  bitmask_type* const __restrict__ destination_mask = out_dcol.null_mask();
```

Comment:
Now that you're using `out_dcol.set_mask_word`, you can remove `destination_mask`

Reply:
Done.